x-pack/plugins/lens/public/indexpattern_datasource/counter_rate/counter_rate.test.ts

```diff
@@ -0,0 +1,392 @@
+          { id: 'split', name: 'split', meta: { type: 'string' } },
+        ],
+        rows: [
+          { val: 1, split: 'A' },
+          { val: 2, split: 'B' },
+          { val: 3 },
+          { val: 4, split: 'A' },
+          { val: 5 },
+          { val: 6, split: 'A' },
+          { val: 7, split: null },
+          { val: 8, split: 'B' },
+          { val: 9, split: '' },
+        ],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output', by: ['split'] }
+    );
+    expect(result.rows.map((row) => row.output)).toEqual([
+      undefined,
+      undefined,
+      undefined,
+      4 - 1,
+      5 - 3,
+      6 - 4,
+      7 - 5,
+      8 - 2,
+      9 - 7,
+    ]);
+  });
+
+  it('calculates counter rate for multiple series by multiple split columns', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [
+          { id: 'val', name: 'val', meta: { type: 'number' } },
+          { id: 'split', name: 'split', meta: { type: 'string' } },
+          { id: 'split2', name: 'split2', meta: { type: 'string' } },
+        ],
+        rows: [
+          { val: 1, split: 'A', split2: 'C' },
+          { val: 2, split: 'B', split2: 'C' },
+          { val: 3, split2: 'C' },
+          { val: 4, split: 'A', split2: 'C' },
+          { val: 5 },
+          { val: 6, split: 'A', split2: 'D' },
+          { val: 7, split: 'B', split2: 'D' },
+          { val: 8, split: 'B', split2: 'D' },
+        ],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output', by: ['split', 'split2'] }
+    );
+    expect(result.rows.map((row) => row.output)).toEqual([
+      undefined,
+      undefined,
+      undefined,
+      4 - 1,
+      undefined,
+      undefined,
+      undefined,
+      8 - 7,
+    ]);
+  });
+
+  it('splits separate series by the string representation of the cell values', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [
+          { id: 'val', name: 'val', meta: { type: 'number' } },
+          { id: 'split', name: 'split', meta: { type: 'string' } },
+        ],
+        rows: [
+          { val: 1, split: { anObj: 3 } },
+          { val: 2, split: { anotherObj: 5 } },
+          { val: 10, split: 5 },
+          { val: 11, split: '5' },
+        ],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output', by: ['split'] }
+    );
+
+    expect(result.rows.map((row) => row.output)).toEqual([undefined, 2 - 1, undefined, 11 - 10]);
+  });
+
+  it('casts values to number before calculating counter rate', () => {
+    const result = runFn(
+      {
+        type: 'datatable',
+        columns: [{ id: 'val', name: 'val', meta: { type: 'number' } }],
+        rows: [{ val: 5 }, { val: '7' }, { val: '3' }, { val: 2 }],
+      },
+      { inputColumnId: 'val', outputColumnId: 'output' }
+    );
+    expect(result.rows.map((row) => row.output)).toEqual([undefined, 7 - 5, 3, 2]);
```

Comment:
Most of the other tests are using numbers that are always increasing, this test appears to be the only one that tests the scenario where the number goes down. Are there other scenarios to test around the edge cases of the algorithm? Or maybe you can name the test differently to indicate that this one tests 2 things?

Reply:
Good idea @wylieconlon! I'll change the name. When it comes to other scenarios, I guess I could test negative values just in case.